ogr/ogrsf_frmts/xodr/ogr_xodr.h

```diff
@@ -0,0 +1,261 @@
+
+    OGRXODRLayerReferenceLine(RoadElements xodrRoadElements,
+                              std::string proj4Defn);
+};
+
+class OGRXODRLayerLaneBorder : public OGRXODRLayer
+{
+  protected:
+    virtual void defineFeatureClass() override;
+    virtual OGRFeature *GetNextFeature() override;
+
+  public:
+    const std::string FEATURE_CLASS_NAME = "LaneBorder";
+
+    OGRXODRLayerLaneBorder(RoadElements xodrRoadElements,
+                           std::string proj4Defn);
+};
+
+class OGRXODRLayerRoadMark : public OGRXODRLayer
+{
+  protected:
+    virtual void defineFeatureClass() override;
+    virtual OGRFeature *GetNextFeature() override;
+
+  public:
+    const std::string FEATURE_CLASS_NAME = "RoadMark";
+
+    OGRXODRLayerRoadMark(RoadElements xodrRoadElements, std::string proj4Defn,
+                         bool dissolveTriangulatedSurface);
+};
+
+class OGRXODRLayerRoadObject : public OGRXODRLayer
+{
+  protected:
+    virtual void defineFeatureClass() override;
+    virtual OGRFeature *GetNextFeature() override;
+
+  public:
+    const std::string FEATURE_CLASS_NAME = "RoadObject";
+
+    OGRXODRLayerRoadObject(RoadElements xodrRoadElements,
+                           std::string proj4Defn);
+};
+
+class OGRXODRLayerRoadSignal : public OGRXODRLayer
+{
+  protected:
+    virtual void defineFeatureClass() override;
+    virtual OGRFeature *GetNextFeature() override;
+
+  public:
+    const std::string FEATURE_CLASS_NAME = "RoadSignal";
+
+    OGRXODRLayerRoadSignal(RoadElements xodrRoadElements, std::string proj4Defn,
+                           bool dissolveTriangulatedSurface);
+};
+
+class OGRXODRLayerLane : public OGRXODRLayer
+{
+  protected:
+    virtual void defineFeatureClass() override;
+    virtual OGRFeature *GetNextFeature() override;
+
+  public:
+    const std::string FEATURE_CLASS_NAME = "Lane";
+
+    OGRXODRLayerLane(RoadElements xodrRoadElements, std::string proj4Defn,
+                     bool dissolveTriangulatedSurface);
+};
+
+/*--------------------------------------------------------------------*/
+/*--------------------  Data source declarations ----------------------*/
+/*--------------------------------------------------------------------*/
+
+class OGRXODRDataSource : public GDALDataset
+{
+  private:
+    OGRXODRLayer **layers;
+    int nLayers;
+
+    /**
+     * Retrieves all necessary road elements from the underlying OpenDRIVE structure.
+     * 
+     * \param roads Roads of the dataset.
+    */
+    RoadElements createRoadElements(const std::vector<odr::Road> roads);
+
+  public:
+    OGRXODRDataSource();
+    ~OGRXODRDataSource();
+
+    /**
+     * Approximation factor for sampling of continuous geometry functions into discrete
+     * OGC Simple Feature geometries.
+    */
+    double eps;
+
+    /**
+     * Whether to dissolve triangulated surfaces which are created from libOpenDRIVE's meshes.
+    */
+    bool dissolveTIN;
+
+    int Open(const char *fileName, char **openOptions, int bUpdate);
```

Comment:
bUpdate likely not needed if you don't support updates.

```suggestion
    bool Open(const char *fileName, CSLConstList openOptions);
```

Reply:
I can't follow why you changed the return value of `Open()` from `int` to `bool`. Should we not be able to return any of the possible different `CPLErr`, such as other drivers do? We will for sure adapt your suggestion but I'd like to better understand the reason and intended driver behaviour :-).

Reply:
There are a ton of legacy use of "int" in GDAL code base where "bool" should have rather been used since the returned values where TRUE or FALSE, and we try to use "bool" whenever we can in new code. I haven't checked here, but I strongly suspect you just need bool. And if you want to return CPLErr then use CPLErr as the return type.

Reply:
Wonderful, that makes total sense. Thanks for the clarification.